worker.go

```diff
@@ -641,3 +642,41 @@ func getCloudWatchAlarmsFromConfigMap(configMap *kubernetes.ConfigMap) aws.Cloud
 
 	return configList
 }
+
+// cniEventHandler syncronizes the events from kubernetes and the status updates from the load balancer controller.
+// Events from kubernetes pod updates trigger a rate limited update of the target groups.
+func cniEventHandler(ctx context.Context, targetCNIcfg *aws.TargetCNIconfig,
+	targetSetter func([]string) error, informer func(context.Context, chan<- []string) error) {
+
+	log.Infoln("Starting CNI event handler")
+
+	rateLimiter := time.NewTicker(cniEventRateLimit)
+	defer rateLimiter.Stop()
+
+	epCh := make(chan []string, 10)
+	go func() {
+		log.Error(informer(ctx, epCh))
+		time.Sleep(cniEventRateLimit)
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case targetCNIcfg.TargetGroupARNs = <-targetCNIcfg.TargetGroupCh:
+			log.Debugf("new message target groups: %v", targetCNIcfg.TargetGroupARNs)
```

Comment:
This is the central event handler that is running once only and receives from channels the events, which is the concurrency solver.
I'm not a fan of locking as it can have hard to track performance issues, but if you want that locking over channels I can refactor.

Reply:
Everything that is triggered by informer should be safe from concurrency pov.

Reply:
It is not really obvious.
targetSetter is indeed `adapter.SetTargetsOnCNITargetGroups` that runs in this loop.
Maybe then we just need to use local variable instead of `targetCNIcfg.TargetGroupARNs` and pass it to the `targetSetter/SetTargetsOnCNITargetGroups` along with endpoints.

Reply:
Informers don't trigger 2 update events at the ~same time, as far as I remember at least.
I think the implementation of informer is something like WATCH, but cache it and have a single goroutine that triggers the callbacks (hello event). If this is true then my statement holds true. So if we do not add a concurrency bug ourselves (spawning a goroutine for example), then it should be safe and that's what I read in the code.
To @AlexanderYastrebov suggestions, makes sense to me.
@universam1 can you check this please?

Reply:
@szuecs Exactly how you described is what I understand and the mere reason I'm in favor of using client-go/SharedInformer/Cache as it solves all those concurrency and eventual consistency for us. My tests hold that true.
Some statements from https://pkg.go.dev/k8s.io/client-go@v0.23.1/tools/cache#SharedInformer
> Every query against the local cache is answered entirely from one snapshot of the cache's state. Thus, the result of a `List` call will not contain two entries with the same namespace and name
> For a given SharedInformer and client, the notifications are delivered sequentially. For a given SharedInformer, client, and object ID, the notifications are delivered in order
> A client must process each notification promptly; a SharedInformer is not engineered to deal well with a large backlog of notifications to deliver. Lengthy processing should be passed off to something else, for example through a `client-go/util/workqueue`.

In our case the chan substitutes.